diff --git pxdvik.orig/dvi-draw.c pxdvik/dvi-draw.c
index f973888..062a944 100644
--- pxdvik.orig/dvi-draw.c
+++ pxdvik/dvi-draw.c
@@ -731,7 +731,7 @@ put_bitmap(struct bitmap *bitmap, int x, int y)
    image) for the given shrink factor. E.g. at shrink level 2, the
    size of the pixel table is 4; in other words, 1 pixel in the shrunk
    image corresponds to 4 pixels in the unshrunk image. Thus, the possible
-   values for shink level 2 are: black, 1/4 (0x404040), 1/2 (0x808080),
+   values for shrink level 2 are: black, 1/4 (0x404040), 1/2 (0x808080),
    3/4 (0xc0c0c0) and white (0xffffff).
 */
 static Pixel *pixeltbl;
@@ -1338,7 +1338,7 @@ shrink_glyph(struct glyph *g)
     }
     g->y2 = g->y / currwin.shrinkfactor;
     if (globals.debug & DBG_BITMAP)
-	print_bitmap(&g->bitmap2);
+        print_bitmap(&g->bitmap2);
 }
 
 #ifdef PTEX
@@ -2252,7 +2252,7 @@ common_set_char(
 	    if (resource.use_grey) {
 		if (g->pixmap2 == NULL) {
 #ifdef DBG_AA
-		    fprintf(stderr, "shinking the bitmap!\n");
+		    fprintf(stderr, "shrinking the bitmap!\n");
 #endif /* DBG_AA */
 		    /*  		    print_bitmap(&g->bitmap); */
 		    shrink_glyph_grey(g);
@@ -2323,7 +2323,7 @@ set_char2(
 	free_bitmap2(g);
 	g->tdir = currinf.fontp->dir;
 	if (globals.debug & DBG_BITMAP) print_char((ubyte) ch, g);
-    }
+   }
     set_char_rotate_glyph(g);
 #ifdef TEXXET
     common_set_char(cmd, g);
diff --git pxdvik.orig/dvi-init.c pxdvik/dvi-init.c
index 58953b7..28d2e06 100644
--- pxdvik.orig/dvi-init.c
+++ pxdvik/dvi-init.c
@@ -229,7 +229,7 @@ reset_fonts(void)
     struct font *f;
 
     for (f = font_head; f != NULL; f = f->next) {
-	if ((f->flags & FONT_LOADED) && !(f->flags & FONT_VIRTUAL)) {
+	if ((f->flags & FONT_LOADED) && !(f->flags & FONT_VIRTUAL)){
 	    delete_glyphs(f);
 	}
     }
@@ -305,6 +305,9 @@ free_unused_fonts(void)
 void
 reset_colors(void)
 {
+#ifdef PTEX
+  int n, maxchar;
+#endif
     if (color_list_len != 0) {
 	XFreeColors(DISP, G_colormap, color_list, color_list_len, 0);
 	color_list_len = 0;
@@ -328,13 +331,18 @@ reset_colors(void)
 	struct glyph *g;
 
 	for (f = font_head; f != NULL; f = f->next)
-	    if ((f->flags & FONT_LOADED) && !(f->flags & FONT_VIRTUAL)
+	  if ((f->flags & FONT_LOADED) && !(f->flags & FONT_VIRTUAL)) {
 #ifdef PTEX
-		&& !(f->flags & FONT_KANJI)
+	    maxchar = (f->flags & FONT_KANJI) ? KTABLESIZE : f->maxchar + 1;
+	    for (n = 0; n < maxchar; ++n) {
+	      g = (f->flags & FONT_KANJI) ? f->kglyph[n] : &f->glyph[n];
+	      if (g == NULL) continue;
+#else
+	      for (g = f->glyph; g <= f->glyph + f->maxchar; ++g) {
 #endif /* PTEX */
-		)
-		for (g = f->glyph; g <= f->glyph + f->maxchar; ++g)
-		    g->fg = NULL;
+		g->fg = NULL;
+	      }
+	  }
     }
 #endif /* GREY */
     bg_current = NULL;
diff --git pxdvik.orig/util.c pxdvik/util.c
index 9188f7d..41d754d 100644
--- pxdvik.orig/util.c
+++ pxdvik/util.c
@@ -212,6 +212,7 @@ struct debug_string_options debug_options[] = {
     {  DBG_FIND,	"find",		", " },
     {  DBG_FILES,	"files",	", " },
     {  DBG_PTEXFNT,	"ptexfnt",	", " },
+    /* {  DBG_AA,		"aa",		", " }, */
     {  DBG_ALL,		"all",		"\n" },
     /* end marker */
     {  0,		NULL,		NULL }
diff --git pxdvik.orig/xdvi-debug.h pxdvik/xdvi-debug.h
index 22feee3..b161ac1 100644
--- pxdvik.orig/xdvi-debug.h
+++ pxdvik/xdvi-debug.h
@@ -102,6 +102,7 @@
 #define DBG_FIND		262144
 #define DBG_FILES		524288
 #define DBG_PTEXFNT             1048576
+/* #define DBG_AA                  2097152 */
 #define	DBG_ALL			(~DBG_BATCH)
 
 /* a mapping of numerical options to descriptive strings, defined in util.c */
